Fix misalignment of table output for multi-byte chars (#268)

chausner opened the pull request:
This PR suggests a fix for #268. The problem is that `std::string::length()` is used in `TableOutput` to determine the length of table cell values. Since `std::string` is not UTF8-aware, it returns the number of bytes, not characters. The two are different if the string contains special characters.

This fix adds a `Utility::UTF8Length` method to obtain the number of characters of a UTF8 string, and uses it in `TableOutput` to compute the display length of cell values.

###### Microsoft Reviewers: [Open in CodeFlow](http://wpcp.azurewebsites.net/CodeFlowProtocolProxy2.php?pullrequest=https://github.com/microsoft/winget-cli/pull/313)

nikeee commented:
If I understand this code correctly, it converts the UTF-8 string into an UTF-16 string and counts the number of wchar_t.
If I'm not mistaking, this would suffer from the same issue as the current implementation, pro problem just occurs on more exotic chars like Chinese or Emojis.
Counting the number of grapheme clusters would be the way to go, I think.
See: https://mortoray.com/2016/04/28/what-is-the-length-of-a-string-a-tricky-question/

chausner commented:
> If I'm not mistaking, this would suffer from the same issue as the current implementation, the problem just occurs on more exotic chars like Chinese or Emojis.

I was wondering about this, too, and you are probably right.

In any case, the way UTF-8 is currently handled in the code base (in plain `std::string`) does not seem ideal because it is very likely that similar mistakes will slip into the code in the future again. C++20 has `std::u8string` but I guess the project cannot use it yet.

JohnMcPMS commented:
Unfortunately u8string won't help either, as @nikeee points out what we really want to count are grapheme clusters.  I was aware of this when I wrote it, but I:
1. Mistakenly thought that it would simply result in a wider column (not thinking that different strings would have different disparities in encoded vs cluster count).
2. Didn't find a simple Windows API to handle it.
3. Ok, I was aware that there were issues, but not to the depths of terminology / semantics around Unicode and characters...

I think that the right thing to do is to implement "character" breakpoint iteration counting from ICU.  Since we use the version of ICU that ships with Windows, that means using the C API as described http://userguide.icu-project.org/boundaryanalysis

In addition to UTF8Length, the TableOutput as written would also need "UTF8Substring" for truncating strings properly.  Also, since we are on the subject, is the ellipses (...) a language independent construct for indicating truncation?

I do appreciate this fix for my haste, but I don't want to sweep it under the rug by fixing only some of the problem.

chausner commented:
> I think that the right thing to do is to implement "character" breakpoint iteration counting from ICU. Since we use the version of ICU that ships with Windows, that means using the C API as described http://userguide.icu-project.org/boundaryanalysis
> 
> In addition to UTF8Length, the TableOutput as written would also need "UTF8Substring" for truncating strings properly.

I changed the implementation to use the code you linked. I also added a `UTF8Substring` method.

Independent from that, in bb92fd9, I changed the three dots into a single ellipsis character. This wastes a little less space and allows to show two more characters of the abbreviated string. But feel free to revert this if you preferred the three dots.

> Also, since we are on the subject, is the ellipses (...) a language independent construct for indicating truncation?

According to [Wikipedia](https://en.wikipedia.org/wiki/Ellipsis), ellipsis seems common in other languages, as well. But some languages use variations of it, e.g. in Chinese they use six instead of three dots.

JohnMcPMS commented:
/azp run

JohnMcPMS commented:
Looks great @chausner 

I would probably move the ICU setup boilerplate into a separate type just so its only written once.  And you can use wil::unique_any_t to define a smart pointer without having to use scope_exits.

Yes, Wikipedia does seem to suggest that it is used in other languages, although it also points to varying ways in which it is represented.  I think that your move to use the single character ellipsis may actually prove to be the right thing with regards to this in any case, both because it shows more characters and because it may just render as something more appropriate in the user's chosen font.

JohnMcPMS commented:
I made the changes that I was suggesting on your branch.  Let me know how you feel about them.

/azp run

JohnMcPMS commented:
/azp run

chausner commented on src/AppInstallerCommonCore/AppInstallerStrings.cpp:
```diff
@@ -12,6 +14,55 @@ namespace AppInstaller::Utility
+                int32_t i = ubrk_first(m_brk.get());
+                if (i != 0)
+                {
+                    AICLI_LOG(Core, Error, << "ubrk_first returned 0");
```
The log message should read "returned non-zero".

chausner commented:
> I made the changes that I was suggesting on your branch. Let me know how you feel about them.

Just the one typo in the log message, otherwise looks good to me. 👍

JohnMcPMS commented:
/azp run

JohnMcPMS commented:
Resolves #268
